packages/minimongo/minimongo.js

```diff
@@ -782,8 +782,22 @@ LocalCollection.prototype.update = function (selector, mod, options, callback) {
   // generate an id for it.
   var insertedId;
   if (updateCount === 0 && options.upsert) {
-    var newDoc = LocalCollection._removeDollarOperators(selector);
+    var newDoc = {};
+     
+    if (LocalCollection._selectorIsId(selector)) {
+      selector = {_id: selector}
+    }
+
+    var selectorDocument = LocalCollection._removeDollarOperators(selector);
+
+    if (selectorDocument._id) {
+      newDoc._id = selectorDocument._id;
+      delete selectorDocument._id;
+    }
+
+    LocalCollection._modify(newDoc, {$set: selectorDocument});
```

Comment:
This double `_modify` call initially through me off. I get why you're making it, and it works, but I think we should add a comment here explaning things a bit for future maintainers. Something along the lines of:

```js
// This double _modify call is made to help work around an issue where collection 
// upserts won't work properly, with nested modifiers (see issue #8631).
```